src/telemetry/metrics/mod.rs

```diff
@@ -209,58 +211,58 @@ impl Root {
-        self.requests.scopes.entry(labels)
+    fn request<F: FnOnce(&mut http::RequestMetrics)>(&self, labels: RequestLabels, f: F) {
+        f(self.requests.lock().expect("lock").scopes.entry(labels)
             .or_insert_with(|| http::RequestMetrics::default().into())
-            .stamped()
+            .stamped());
     }
 
-    fn response(&mut self, labels: ResponseLabels) -> &mut http::ResponseMetrics {
-        self.responses.scopes.entry(labels)
+    fn response<F: FnOnce(&mut http::ResponseMetrics)>(&self, labels: ResponseLabels, f: F) {
+        f(self.responses.lock().expect("lock").scopes.entry(labels)
             .or_insert_with(|| http::ResponseMetrics::default().into())
-            .stamped()
+            .stamped());
     }
 
-    fn transport(&mut self, labels: TransportLabels) -> &mut transport::OpenMetrics {
-        self.transports.scopes.entry(labels)
+    fn transport<F: FnOnce(&mut transport::OpenMetrics)>(&self, labels: TransportLabels, f: F) {
+        f(self.transports.lock().expect("lock").scopes.entry(labels)
             .or_insert_with(|| transport::OpenMetrics::default().into())
-            .stamped()
+            .stamped());
     }
 
-    fn transport_close(&mut self, labels: TransportCloseLabels) -> &mut transport::CloseMetrics {
-        self.transport_closes.scopes.entry(labels)
+    fn transport_close<F: FnOnce(&mut transport::CloseMetrics)>(&self, labels: TransportCloseLabels, f: F) {
+        f(self.transport_closes.lock().expect("lock").scopes.entry(labels)
             .or_insert_with(|| transport::CloseMetrics::default().into())
-            .stamped()
+            .stamped())
     }
 
-    fn tls_config(&mut self, labels: TlsConfigLabels) -> &mut Counter {
-        self.tls_config.scopes.entry(labels)
-            .or_insert_with(|| Counter::default())
+    fn tls_config<F: FnOnce(&mut Counter)>(&self, labels: TlsConfigLabels, f: F) {
+        f(self.tls_config.lock().expect("lock").scopes.entry(labels)
+            .or_insert_with(|| Counter::default()))
     }
 
-    fn retain_since(&mut self, epoch: Instant) {
-        self.requests.retain_since(epoch);
-        self.responses.retain_since(epoch);
-        self.transports.retain_since(epoch);
-        self.transport_closes.retain_since(epoch);
+    fn retain_since(&self, epoch: Instant) {
+        self.requests.lock().expect("lock").retain_since(epoch);
+        self.responses.lock().expect("lock").retain_since(epoch);
+        self.transports.lock().expect("lock").retain_since(epoch);
+        self.transport_closes.lock().expect("lock").retain_since(epoch);
     }
 }
 
 impl fmt::Display for Root {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        self.requests.fmt(f)?;
-        self.responses.fmt(f)?;
-        self.transports.fmt(f)?;
-        self.transport_closes.fmt(f)?;
-        self.tls_config.fmt(f)?;
+        self.requests.lock().expect("lock").fmt(f)?;
```

Comment:
I wonder if the `self.$THING.lock().expect("lock")` could be made less repetitive...maybe a macro?

Reply:
As this is incremental, I haven't been very focused on polishing any of this, as it will all change in followups

Reply:
Yeah, I figured. Definitely not a blocker.